exploration/number-selection.md

```diff
@@ -92,44 +95,274 @@ ICU MF1 messages using `plural` and `selectordinal` should be representable in M
+   - Else, if the value of `key` is a keyword:
+      - Let `keyword` be a string which is the result of [rule selection](#rule-selection).
+      - If `keyword` equals `key`, then `key` matches the selector.
+        Append `key` to the end of the `return_value` list.
+   - Else, `key` is not a match.
+     Do not add `key` to `return_value`
+- Return `return_value`
+
+### Plural/Ordinal Keywords
+The _plural/ordinal keywords_ are: `zero`, `one`, `two`, `few`, `many`, and
+`other`.
+The value `other` is both a valid output of [rule selection](#rule-selection)
+and serves a default value.
+
+### Rule Selection
+
+If the option `select` is set to `exact`, rule-based selection is not used.
+Return the empty string.
+
+> [!NOTE]
+> Since keys cannot be the empty string in a numeric selector, returning the
+> empty string disables keyword selection
+
+If the option `select` is set to `plural`, selection should be based on CLDR plural rule data
+of type `cardinal`. See [charts](https://www.unicode.org/cldr/charts/latest/supplemental/language_plural_rules.html)
+for examples.
+
+If the option `select` is set to `ordinal`, selection should be based on CLDR plural rule data
+of type `ordinal`. See [charts](https://www.unicode.org/cldr/charts/latest/supplemental/language_plural_rules.html)
+for examples.
+
+Apply the rules defined by CLDR to the resolved value of the operand and return the resulting keyword.
+If no rules match, return `other`.
+
+> **Example.**
+> In CLDR 44, the Czech (`cs`) plural rule set can be found
+> [here](https://www.unicode.org/cldr/charts/44/supplemental/language_plural_rules.html#cs).
+>
+> A message in Czech might be:
+> ```
+> .match {$numDays :number}
+> one  {{{$numDays} den}}
+> few  {{{$numDays} dny}}
+> many {{{$numDays} dne}}
+> *    {{{$numDays} dní}}
+> ```
+> Using the rules found above, the results of various `operand` values might look like:
+> | Operand value | Keyword | Formatted Message |
+> |---|---|---|
+> | 1 | `one` | 1 den |
+> | 2 | `few` | 2 dny |
+> | 5 | `other` | 5 dní |
+> | 22 | `few` | 22 dny |
+> | 27 | `other` | 27 dní |
+> | 2.4 | `many` | 2,4 dne |
 
-If PR [#532](https://github.com/unicode-org/message-format-wg/pull/532) is accepted,
-also add the following `<alias>` definitions to `<function name="number">`:
 
-```xml
-<alias name="plural" supports="match">
-	<setOption name="select" value="plural"/>
-</alias>
-<alias name="ordinal" supports="match">
-	<setOption name="select" value="ordinal"/>
-</alias>
-```
+
+### Determining Exact Literal Match
+
+Number literals in the MessageFormat v2 syntax use the 
+[format defined for a JSON number](https://www.rfc-editor.org/rfc/rfc8259#section-6).
+The resolved value of an `operand` exactly matches a numeric literal `key`
+if, when the `operand` is serialized using the format for a JSON number
+including the exact number of fraction digits specified by the selector
+function or its options, the two strings are equal.
```

Comment:
This places an unreasonable burden on all selections in order to support a theoretical corner case. Consider, for instance, that we also support the options `minimumIntegerDigits`, `minimumSignificantDigits`, and `maximumSignificantDigits`. They either need to be considered here as well, or ignored.

I'm starting to think we don't need to specify this here, and should leave it up to implementations. Comparisons against floating point values are fraught in general, and should be avoided.

```suggestion
The resolved value of an `operand` exactly matches a numeric literal `key`
if, when the `operand` is serialized using the format for a JSON number,
the two strings are equal.
All formatting options are ignored when applying this serialization.
Using keys representing fractional numbers is not recommended,
and may produce different results with different implementations.
```

If this suggestion is accepted, the examples below will also need to be updated.

Reply:
Integer and significant digits (and, fwiw, grouping) formatting specifications don't enter into it because they don't affect plural selection the way that fraction digits do. They surely affect the appearance of the final product, but not "equality".

This means that overlong keys probably don't work:

```
.match {$num :number minimumIntegerDigits=3}
000 {{I can never be selected}}
010 {{I can't either}}
100 {{But I work}}
```

> Comparisons against floating point values are fraught in general, and should be avoided.

Yes. I'll note that people using this feature are on thin ice (this is actually the one place where `ChoiceFormat` did something useful in the modern era). Nearly all exact number selections that I'm familiar with use integers.

Reply:
Ok, I'll grant you that `minimumIntegerDigits` can be discounted, but:

```js
new Intl.NumberFormat('en', { maximumSignificantDigits: 2 }).format(1.03) // "1"
new Intl.NumberFormat('en', { maximumSignificantDigits: 3 }).format(1.03) // "1.03"

new Intl.NumberFormat('en', { minimumSignificantDigits: 2 }).format(1.3) // "1.3" 
new Intl.NumberFormat('en', { minimumSignificantDigits: 3 }).format(1.3) // "1.30"
```

Reply:
> This places an unreasonable burden on all selections in order to support a theoretical corner case

I disagree. The suggested language for this will break for some cases, and not just theoretical ones.

The language "Apply the rules defined by CLDR to the resolved value of the operand and the function options,
and return the resulting keyword." works, and is not an excessive burden. Worst case is that as well as formatting the variable, you also format a few literals. However, it is very easy to optimize the implementations to avoid the extra formatting.

Reply:
@eemeli Fair enough. I added significant digits to the match, including examples.

I do notice that both FF and Chrome's implementation of `Intl.NumberFormat` prioritizes significant digits ahead of fraction ones:

```javascript
const number = 1.23456;

console.log(
  new Intl.NumberFormat('de-DE', { maximumSignificantDigits:3, maximumFractionDigits:0 }).format(
    number,
  ),
);
```

Produces: `1,23` 

(ICU4J `NumberFormatter` does not replicate this, so this might be a bug)

```java
        NumberFormatter.withLocale(Locale.US)
           .precision(Precision.minSignificantDigits(3))
           .precision(Precision.maxFraction(0))
           .format(1.2345)
           .toString()  // produces `1`
       );
```

In any case, I didn't deal with the interplay between these two in the spec and \* _sigh_ \* I probably need to.

Reply:
@sffc is the expert on this. The best interplay between them where there are seeming conflicts is tricky to get right, and as I recall, _not_ just a matter of having one have precedence over the other.

Reply:
My previous research showed that min/max significant digits are much less popular than min/max fraction digits outside of compact notation. If you don't need compact notation, I would suggest considering dropping significant digits.

There is a way to map between ECMA-402 digit options and NumberFormatter digit options. I maintain that the NumberFormatter options are simpler and less error-prone, but the ECMA-402 options were set long ago so we did the best we could within that space.

Reply:
Thanks @sffc. Significant digits are definitely going to be rarer than fraction digits--I can't recall using them in any kind of "normal" application, aside from compact (and then only sporadically). 

I have no problem with ECMA-402's options. I just note that it produced some surprise for me to get a fractional part after I explicitly asked not to (e.g. significant > fraction). I expected that significant digits would step on the *value* of the formattable number, while fraction digits stepped on the number of decimal places shown.

We *only* have those options that overlap between 402 and ICU4J (so, for example, no rounding control) and since significant digits is in both, that's why it's there. I would shed zero tears to drop it in preview, although it's just going to come back again later (or in the form of `:icu:maximumSignificantDigits`/`:js:maximumSignficantDigits` 😉)

In any case, we have differences in implementations, so I should tread carefully in MFv2 and make the interplay implementation defined (assuming you do not agree that ECMA-402's behavior is a bug).

Reply:
> We only have those options that overlap between 402 and ICU4J (so, for example, no rounding control)

ICU4J is a superset of ECMA-402. Note that ECMA-402 contains rounding increment and rounding mode as of NFv3 which was stabilized about a year ago.

> In any case, we have differences in implementations, so I should tread carefully in MFv2 and make the interplay implementation defined

I don't agree with the characterization that there is a "difference in implementations". Rather, there is a difference in APIs. It should not be assumed that just because there is an option with the same / similar name in ECMA-402 and ICU4J that the two options have the same / similar behavior. When the APIs were designed (I was involved with both), the respective committees preferred to learn from the mistakes of the previous generation than try to match it bug-for-bug.